File: java/src/main/java/com/mycompany/pt/PaymentListing.java
/*
 * |-------------------------------------------------
 * | Copyright © 2017 Colin But. All rights reserved.
 * |-------------------------------------------------
 */
package com.mycompany.pt;

import com.mycompany.pt.model.CurrencyCode;
import com.mycompany.pt.model.MoneyAmount;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class PaymentListing {

    private static PaymentListing INSTANCE = new PaymentListing();

    private Map<CurrencyCode, MoneyAmount> paymentAmounts = new HashMap<>();

    public static PaymentListing getInstance() {
        return INSTANCE;
    }

    public Map<CurrencyCode, MoneyAmount> getPaymentAmounts() {
        return paymentAmounts;
    }

    public void addPayment(CurrencyCode currencyCode, BigDecimal amount) {
        if (paymentAmounts.get(currencyCode) != null) {
            MoneyAmount existingPaymentAmount = paymentAmounts.get(currencyCode);
            existingPaymentAmount.setAmount(existingPaymentAmount.getAmount().add(amount));
        } else {
            paymentAmounts.put(currencyCode, new MoneyAmount(currencyCode, amount));
        }
    }
}

File: kotlin/src/main/kotlin/payment/tracker/PaymentListing.kt
/*
 * |-------------------------------------------------
 * | Copyright © 2019 Colin But. All rights reserved.
 * |-------------------------------------------------
 */
package payment.tracker

import payment.tracker.model.CurrencyCode
import payment.tracker.model.MoneyAmount
import java.math.BigDecimal

// TODO convert to Singleton
class PaymentListing (val paymentAmounts : MutableMap<CurrencyCode, MoneyAmount>) {

    fun addPayment(currencyCode: CurrencyCode, amount: BigDecimal) {
        if (paymentAmounts[currencyCode] != null) {
            val existingPayment : MoneyAmount = paymentAmounts[currencyCode]!!
            existingPayment.amount = existingPayment.amount.add(amount)
        } else {
            paymentAmounts.putIfAbsent(currencyCode, MoneyAmount(currencyCode, amount))
        }
    }
}
